File: javascript_basics/src/tests/form_auth.test.js
require("../testSetup.js");
const { By, until } = require("selenium-webdriver");
const { getDriver } = require("../globalSetup.js");
const { logger } = require("../utils/logger.js");
const { sleep } = require("../utils/sleep.js");

describe("Form Authentication", function () {

  //
  // FIRST TEST — INVALID LOGIN
  //
  it("should fail to log in with invalid credentials", async function () {

    const driver = getDriver();
    logger.info("STARTING Test --- Failed Login ---");

    const url = "https://the-internet.herokuapp.com/login";
    logger.info(`Navigating to ${url}`);
    await driver.get(url);
    // await sleep(1000);

    logger.info("Locating username field");
    const usernameField = await driver.findElement(By.id("username"));
    // await sleep(1000);

    logger.info("Entering invalid username");
    await usernameField.sendKeys("wronguser");
    // await sleep(1000);

    logger.info("Locating password field");
    const passwordField = await driver.findElement(By.id("password"));
    // await sleep(1000);

    logger.info("Entering invalid password");
    await passwordField.sendKeys("wrongpassword");
    // await sleep(1000);

    logger.info("Locating Login button");
    const loginButton = await driver.findElement(By.css("button[type='submit']"));
    // await sleep(1000);

    logger.info("Clicking Login button");
    await loginButton.click();
    // await sleep(1000);

    logger.info("Waiting for failure message");
    const flash = await driver.wait(until.elementLocated(By.id("flash")), 5000);
    const message = await flash.getText();
    // await sleep(1000);

    if (
      !message.includes("Your username is invalid!") &&
      !message.includes("Your password is invalid!")
    ) {
      throw new Error(`Expected failure message but got: ${message}`);
    }

    logger.info("ENDING Test --- Failed Login ---");
  });



  //
  // SECOND TEST — VALID LOGIN
  //
  it("should successfully log in with valid credentials", async function () {

    const driver = getDriver();
    logger.info("STARTING Test --- Successful Login ---");

    const url = "https://the-internet.herokuapp.com/login";
    logger.info(`Navigating to ${url}`);
    await driver.get(url);
    // await sleep(1000);

    logger.info("Locating username field");
    const usernameField = await driver.findElement(By.id("username"));
    // await sleep(1000);

    logger.info("Entering username");
    await usernameField.sendKeys("tomsmith");
    // await sleep(1000);

    logger.info("Locating password field");
    const passwordField = await driver.findElement(By.id("password"));
    // await sleep(1000);

    logger.info("Entering password");
    await passwordField.sendKeys("SuperSecretPassword!");
    // await sleep(1000);

    logger.info("Locating Login button");
    const loginButton = await driver.findElement(By.css("button[type='submit']"));
    // await sleep(1000);

    logger.info("Clicking Login button");
    await loginButton.click();
    // await sleep(1000);

    logger.info("Waiting for success message");
    const flash = await driver.wait(until.elementLocated(By.id("flash")), 5000);
    const message = await flash.getText();
    // await sleep(1000);

    if (!message.includes("You logged into a secure area!")) {
      throw new Error(`Expected success message but got: ${message}`);
    }

    logger.info("ENDING Test --- Successful Login ---");
  });

});

File: csharp_basics/tests/form_auth_test.cs
using System;
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using csharp_basics.utils;

namespace csharp_basics.tests
{
    [TestFixture]
    public class FormAuthTest : BaseTest
    {
        [Test]
        public void TestFormAuthenticationValidAndInvalid()
        {
            Logger.Write("STARTING Test --- Form Authentication ---");

            string baseUrl = "https://the-internet.herokuapp.com/login";
            driver.Navigate().GoToUrl(baseUrl);
            Logger.Write($"Navigated to {baseUrl}");

            // Invalid login
            Logger.Write("Attempting invalid login...");
            driver.FindElement(By.Id("username")).SendKeys("invalidUser");
            driver.FindElement(By.Id("password")).SendKeys("wrongPass");
            driver.FindElement(By.CssSelector("button[type='submit']")).Click();

            string errorMsg = driver.FindElement(By.Id("flash")).Text;
            Logger.Write($"Invalid login message: {errorMsg}");
            Assert.That(errorMsg, Does.Contain("Your username is invalid!"));

            // Valid login
            Logger.Write("Attempting valid login...");
            driver.Navigate().GoToUrl(baseUrl);
            driver.FindElement(By.Id("username")).SendKeys("tomsmith");
            driver.FindElement(By.Id("password")).SendKeys("SuperSecretPassword!");
            driver.FindElement(By.CssSelector("button[type='submit']")).Click();

            var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(5));
            Logger.Write("Waiting for success message to appear...");

            IWebElement successElement = wait.Until(d =>
            {
                try
                {
                    var el = d.FindElement(By.Id("flash"));
                    return el.Displayed ? el : null;
                }
                catch (NoSuchElementException)
                {
                    return null;
                }
            });

            string successMsg = successElement.Text;
            Logger.Write($"Valid login message: {successMsg}");
            Assert.That(successMsg, Does.Contain("You logged into a secure area!"));

            Logger.Write("ENDING Test --- Completed successfully: Form Authentication ---");
        }
    }
}
